Add a new build factory to create the staged builders (StagedBuilder).

vvereschaka opened the pull request:
* StagedBuilder.getCmakeBuildFactory

The factory provides a simple staged builder configuration to build the LLVM project unified tree.

This factory is a replacement for `UnifiedTreeBuilder.getCmakeWithNinjaMultistageBuildFactory` factory and it also can replace at least `FlangBuilder.getFlangOutOfTreeBuildFactory`.

vvereschaka commented:
An example of the builder configuration with this build factory can be found here https://github.com/llvm/llvm-zorg/pull/42

petrhosek commented:
This seems to be replicating a lot of the existing logic that already exists inside the LLVM bootstrapping and runtimes build. Would it be possible to use the existing build support (improving it where needed) instead?

vvereschaka commented:
This factory actually a replacement for `UnifiedTreeBuilder.getCmakeWithNinjaMultistageBuildFactory`, `FlangBuilder.getFlangOutOfTreeBuildFactory`, flang tuntime and any other factories for the **simple** staged builders.
The clang/bootstrap build factories are too specific and complicated currently and they cannot be used for the simple staged builds.

gkistanova commented on zorg/buildbot/builders/StagedBuilder.py:
```diff
@@ -0,0 +1,299 @@
+            rf.enable_runtimes = frozenset(rf.enable_runtimes.union(sf.enable_runtimes))
+            rf.depends_on_projects = set(rf.depends_on_projects.union(sf.enable_runtimes))
+        # Update a list of enabled projects for the root factory.
+        rf.enable_projects = f.depends_on_projects.difference(f.enable_runtimes)
```
Why does it depend on `f`?

gkistanova reviewed (changes requested):
Thanks for the patch, Vlad!

Could you address a few things, please?

Plus I feel like I must be missing something with your `normalize_factory`, thus a few questions.

vvereschaka commented:
@gkistanova I have updated this PR with the following changes (in additional to your change requests):
* fixed step names;
* guards for the max step names (48 characters +2 for BB identical name counters) (db limitation);
* multiple install targets for each stage;
* added properly formatted function's help/description;

gkistanova reviewed (approved):
LGTM.

slydiman reviewed (approved):
LGTM
